### PR TITLE
feat(sdk-coin-avaxp): implement export tx builder

### PR DESCRIPTION
**Pull request by jrafael-simtlix**
[BG-55035](https://bitgoinc.atlassian.net/browse/BG-55035)

and import tx builder
[BG-55125](https://bitgoinc.atlassian.net/browse/BG-55125)

### CODE REVIEW

File: modules/sdk-coin-avaxp/src/lib/transaction.ts [hunk 1]
```diff
@@ -68,6 +68,7 @@ export class Transaction extends BaseTransaction {
   public _fromAddresses: BufferAvax[] = [];
   public _rewardAddresses: BufferAvax[];
   public _utxos: DecodedUtxoObj[] = [];
+  public _fee: BN = new BN(0);
```
Comment (jicli) [inline modules/sdk-coin-avaxp/src/lib/transaction.ts:71]: fee is not 0 in terms of export and import

**Comment by jicli** [comment]
Can you please run a test on `addValidator` and `addDelegator` to ensure we dont have regression?

File: modules/sdk-coin-avaxp/src/lib/delegatorTxBuilder.ts [hunk 2]
```diff
@@ -150,16 +145,16 @@ export class DelegatorTxBuilder extends TransactionBuilder {
     this._startTime = baseTx.getStartTime();
     this._endTime = baseTx.getEndTime();
     this._stakeAmount = baseTx.getStakeAmount();
-    this.transaction._utxos = this.recoverUtxos(baseTx.getIns());
```
Comment (jicli) [inline modules/sdk-coin-avaxp/src/lib/delegatorTxBuilder.ts:153]: do we not need any utxos?

File: modules/sdk-coin-avaxp/src/lib/transferBuilder.ts [hunk 3]
```diff
@@ -10,7 +10,7 @@ export class TransferBuilder extends TransactionBuilder {
   }
 
   protected get transactionType(): TransactionType {
-    return TransactionType.WalletInitialization;
+    return TransactionType.Send;
```
Comment (jicli) [inline modules/sdk-coin-avaxp/src/lib/transferBuilder.ts:13]: there's no send in avaxp ?

File: modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  initBuilder(tx: Tx): this {
+    super.initBuilder(tx);
+    const baseTx: BaseTx = tx.getUnsignedTx().getTransaction();
+    if (!this.verifyTxType(baseTx)) {
+      throw new NotSupported('Transaction cannot be parsed or has an unsupported transaction type');
+    }
+    this._targetChainId = baseTx.getDestinationChain();
+    this._amount = (baseTx.getExportOutputs()[0].getOutput() as any as AmountOutput).getAmount();
```
Comment (jicli) [inline modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts:79]: why do we only take the first output's amount?

File: modules/sdk-coin-avaxp/src/lib/transactionBuilder.ts [hunk 5]
```diff
@@ -148,6 +156,172 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
...
+    // get outputs and credentials from the deserialized transaction if we are in OVC
+    return {
+      inputs,
+      outputs: outputs.length === 0 ? this.transaction.avaxPTransaction.getOuts() : outputs,
+      credentials: credentials.length === 0 ? this.transaction.credentials : credentials,
+    };
+  }
+
+  /**
+   * Inputs can be controlled but outputs get reordered in transactions
```
Comment (jicli) [inline modules/sdk-coin-avaxp/src/lib/transactionBuilder.ts:301]: does the comment match what it's doing? 
`this.transaction._utxos = this.recoverUtxos(baseTx.getImportInputs());` 
doesnt this mean input's not controlled?

File: modules/sdk-coin-avaxp/src/lib/importTxBuilder.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  initBuilder(tx: Tx): this {
+    super.initBuilder(tx);
+    const baseTx: BaseTx = tx.getUnsignedTx().getTransaction();
+    if (!this.verifyTxType(baseTx)) {
+      throw new NotSupported('Transaction cannot be parsed or has an unsupported transaction type');
+    }
+    this._targetChainId = baseTx.getSourceChain();
+    this.transaction._utxos = this.recoverUtxos(baseTx.getImportInputs());
```
Comment (jicli) [inline modules/sdk-coin-avaxp/src/lib/importTxBuilder.ts:48]: why do we override the utxos in import but not export?

File: modules/sdk-coin-avaxp/src/lib/transaction.ts (hunk 1 above)
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/transaction.ts:71]: That is the default fee for AddValidator and AddDelegator txs. 
Should I remove default value and set fee in AddDelegator?

File: modules/sdk-coin-avaxp/src/lib/transferBuilder.ts (hunk 3 above)
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/transferBuilder.ts:13]: The transferBuilder is there. I fix its transactionType, Should I remove it directly?

File: modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts (hunk 4 above)
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts:79]: Ok, I could review full array, but I should be only one, if it's builded by SDK

File: modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    this.transaction._fee = new BN(this.transaction._network.txFee);
+  }
+
+  protected get transactionType(): TransactionType {
+    return TransactionType.Export;
+  }
+
+  /**
+   *
+   * @param value
```
Comment (CapnMigraine) [inline modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts:32]: nit: fix param

File: modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    if (amount.lten(0)) {
+      throw new BuildTransactionError('Amount must be greater than 0');
+    }
+  }
+
+  /**
+   *
+   * @param chainID
+   */
+  private validateChainId(chainID: BufferAvax) {
```
Comment (CapnMigraine) [inline modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts:66]: Is there a reason to make this method private?
This method and `validateAmount` could also be static
```suggestion
  static validateChainId(chainID: BufferAvax): void {
```

File: modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts (hunk 4 above)
Comment (CapnMigraine) [reply modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts:79]: If we have more than 1 output we should either count them or throw an exception if we only expect 1

File: modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    this._targetChainId = baseTx.getDestinationChain();
+    this._amount = (baseTx.getExportOutputs()[0].getOutput() as any as AmountOutput).getAmount();
+    return this;
+  }
+
+  static get txType(): number {
+    return PlatformVMConstants.EXPORTTX;
+  }
+
+  verifyTxType(baseTx: BaseTx): baseTx is ExportTx {
```
Comment (CapnMigraine) [inline modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts:87]: this method could be static

File: modules/sdk-coin-avaxp/src/lib/importTxBuilder.ts [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    this.transaction._fee = new BN(this.transaction._network.txFee);
+  }
+
+  protected get transactionType(): TransactionType {
+    return TransactionType.Import;
+  }
+
+  /**
+   *
+   * @param value
```
Comment (CapnMigraine) [inline modules/sdk-coin-avaxp/src/lib/importTxBuilder.ts:22]: nit: fix param

File: modules/sdk-coin-avaxp/src/lib/importTxBuilder.ts [hunk 11]
```diff
@@ -0,0 +1,83 @@
+import { BaseCoin as CoinConfig } from '@bitgo/statics';
+import { BuildTransactionError, NotSupported, TransactionType } from '@bitgo/sdk-core';
+import { TransactionBuilder } from './transactionBuilder';
+import { BaseTx, ImportTx, PlatformVMConstants, Tx, UnsignedTx } from 'avalanche/dist/apis/platformvm';
+import utils from './utils';
+import { BN, Buffer as BufferAvax } from 'avalanche';
+
+export class ImportTxBuilder extends TransactionBuilder {
```
Comment (CapnMigraine) [inline modules/sdk-coin-avaxp/src/lib/importTxBuilder.ts:8]: Most of the methods are dupes of the ones in the exportTxBuilder.

Can we have an `AtomicTransactionBuilder` that extends `TransactionBuilder` and it's extended by `exportTxBuilder` and `ImportTxBuilder`?

File: modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+  protected get transactionType(): TransactionType {
+    return TransactionType.Export;
+  }
+
+  /**
+   *
+   * @param value
+   */
```
Comment (joey-bai) [inline modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts:30]: fill out js doc

File: modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+
+  /**
+   *
+   * @protected
+   */
+  protected buildAvaxpTransaction(): void {
+    const { inputs, outputs, credentials } = this.createInputOutput(this._amount.add(this.transaction._fee));
+    this.transaction.setTransaction(
+      new Tx(
+        new UnsignedTx(
```
Comment (joey-bai) [inline modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts:69]: if you serialize a signed tx, then de-serialize it, it seems like you'll get an unsigned tx back.

is that intentional?

File: modules/sdk-coin-avaxp/src/lib/transactionBuilder.ts [hunk 14]
```diff
@@ -132,7 +99,72 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
...
+    return this;
+  }
+
+  threshold(value: number): this {
+    this.validateThreshold(value);
+    this._transaction._threshold = value;
+    return this;
+  }
+
+  locktime(value: string | number): this {
```
Comment (joey-bai) [inline modules/sdk-coin-avaxp/src/lib/transactionBuilder.ts:130]: for instance, is the value in seconds? microseconds? nano?
is it okay to have decimals?

File: modules/sdk-coin-avaxp/src/lib/utils.ts [hunk 15]
```diff
@@ -93,7 +93,7 @@ export class Utils implements BaseUtils {
       return false;
     }
   }
-  public parseAddress = (pub: string): BufferAvax => this.binTools.parseAddress(pub, 'P');
+  public parseAddress = (pub: string): BufferAvax => this.binTools.stringToAddress(pub);
```
Comment (jicli) [inline modules/sdk-coin-avaxp/src/lib/utils.ts:96]: if we remove the 'p' it's not gonna work, is it?

File: modules/sdk-coin-avaxp/src/lib/importTxBuilder.ts [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+  protected buildAvaxpTransaction(): void {
+    const { inputs, outputs, credentials } = this.createInputOutput(this.transaction._fee);
+    this.transaction.setTransaction(
+      new Tx(
+        new UnsignedTx(
+          new ImportTx(
+            this.transaction._networkID,
+            this.transaction._blockchainID,
+            outputs,
+            [], // P-Chain input
```
Comment (jicli) [inline modules/sdk-coin-avaxp/src/lib/importTxBuilder.ts:60]: sry, but why is this empty?

File: modules/sdk-coin-avaxp/src/lib/atomicTransactionBuilder.ts [hunk 17]
```diff
@@ -0,0 +1,30 @@
+import { BaseCoin as CoinConfig } from '@bitgo/statics';
+import { BN, Buffer as BufferAvax } from 'avalanche';
+import utils from './utils';
+import { TransactionBuilder } from './transactionBuilder';
+
+/**
+ * Atomic transaction are the cross chain transaction.
+ */
+export abstract class AtomicTransactionBuilder extends TransactionBuilder {
+  private static _cChainId: BufferAvax = utils.cb58Decode('yH8D7ThNJkxmtkuv2jgBa4P1Rn3Qpr4pPr7QYNfcdoS6k6HWp');
```
[the remark this answers is not in the document]
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/atomicTransactionBuilder.ts:10]: Yes, there are. 

In the case of Avaxp, they are added in static
Mainnet:
https://github.com/BitGo/BitGoJS/blob/d76a727737f2fa4e37a253801dcaaa6013882147/modules/statics/src/networks.ts#L175
Testnet:
https://github.com/BitGo/BitGoJS/blob/d76a727737f2fa4e37a253801dcaaa6013882147/modules/statics/src/networks.ts#L200

File: modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts (hunk 13 above)
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts:69]: You are right. Now, I avoid that case
https://github.com/BitGo/BitGoJS/blob/7df888eca68bbc60505d837cdf8729c69f4d387e/modules/sdk-coin-avaxp/src/lib/exportTxBuilder.ts#L78-L79

File: modules/sdk-coin-avaxp/src/lib/importTxBuilder.ts (hunk 16 above)
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/importTxBuilder.ts:60]: The input is from P chain.
The Import input is lower

**Pull request closed without merging**